File: languages/montithings/src/main/java/montithings/trafos/patterns/AutoregressiveAnomalyDetection.java
package montithings.trafos.patterns;

import java.util.*;

/**
 * One of the most basic stochastical models for univariate time-series is the Autoregressive model (AR). AR is a
 * linear model where current value Xt of the stochastic process (dependent variable) is based one a finite set of
 * previous values (independent variables) of length p and an error value e:
 */
public class AutoregressiveAnomalyDetection extends LinearRegression implements UnivariateAnomalyDetection{

    private final int windowSize;
    private final double tolerance;
    private final List<Double> pastValues;

    public AutoregressiveAnomalyDetection(int windowSize, double tolerance) {
        this.windowSize = windowSize;
        this.tolerance = tolerance;
        this.pastValues = new ArrayList<>();
    }

    @Override
    public boolean isAnomaly(Double input) {
        List<Double> regressionValues = this.getRegressionValues();

        this.pastValues.add(input);

        Double predictedValue = this.predictValue(input, regressionValues);

        if (predictedValue == null) {
            return false;
        }

        double err = Math.abs(predictedValue - input);

        return err > this.tolerance;
    }

    private List<Double> getRegressionValues() {
        List<Double> values = new ArrayList<>();

        ListIterator<Double> li = this.pastValues.listIterator(this.pastValues.size());

        while (li.hasPrevious() && values.size() < this.windowSize) {
            values.add(li.previous());
        }

        return values;
    }
}

File: generators/montithings2cpp/src/main/java/montithings/generator/steps/trafos/patterns/hwc/AutoregressiveAnomalyDetection.cpp
#include "AutoregressiveAnomalyDetection.h"

bool AutoregressiveAnomalyDetection::is_anomaly(float input)
{
  std::vector<float> regression_values = this->get_regression_values();

  this->past_values.push_back(input);

  float predicted_value = this->lr->predict_value(input, regression_values);

  float err = std::abs(predicted_value - input);

  return err > this->tolerance;
}

std::vector<float> AutoregressiveAnomalyDetection::get_regression_values()
{
  int values_len = std::min({this->window_size, static_cast<int>(this->past_values.size())});

  std::vector<float> values(values_len);

  for (int i = 0; i < values_len; i++)
  {
    values.push_back(this->past_values[i]);
  }

  return values;
}